File: C#/RendezvousChannelSynchronizer/RendezvousChannelSynchronizer/TestRendezvousChannel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace RendezvousChannelSynchronizer
{
    class TestRendezvousChannel
    {
        //Responses -> Client gives an integer, server gives back the symmetric of it 
        static void Main(string[] args)
        {
            TestRendezvousChannel tests = new TestRendezvousChannel();
            
            //tests.Test_Client_Starts_First_Rendezvous();
            //tests.Test_Server_Starts_First_Rendezvous();
            //tests.Test_Client_Zero_Time_Rendezvous();
            //tests.Test_Client_Zero_Time_After_Server_Starts_Rendezvous();
            //tests.Test_Server_Zero_Time_Rendezvous();
            //tests.Test_Server_Zero_Time_After_CLient_Starts_Rendezvous();
            //tests.Test_Multiple_Client_One_Server_Rendezvous();
            //tests.Test_Multiple_Server_One_Client_Rendezvous();
            //tests.Test_Multiple_Clients_First_And_Servers();
            //tests.Test_Multiple_Servers_First_And_Clients();
            //tests.Test_Long_Wait_In_Rendezvous();
            //tests.Test_Client_Interrupted_Rendezvous();
            //tests.Test_Server_Interrupted_Rendezvous();

            Console.ReadKey();
        }

        private void Test_Client_Starts_First_Rendezvous()
        {
            RendezvousChannel<int, int> rendChannel = new RendezvousChannel<int, int>();
            Thread t1 = new Thread(() => ServerFunc(1, rendChannel, 5000, 500));
            Thread t2 = new Thread(() => ClientFunc(1, rendChannel, 10, 5000));

            t1.Start();
            Thread.Sleep(500);
            t2.Start();
            Thread.Sleep(500);
        }

        private void Test_Server_Starts_First_Rendezvous()
        {
            RendezvousChannel<int, int> rendChannel = new RendezvousChannel<int, int>();
            Thread t1 = new Thread(() => ClientFunc(1, rendChannel, 30, 5000));
            Thread t2 = new Thread(() => ServerFunc(1, rendChannel, 5000, 500));

            t1.Start();
            Thread.Sleep(500);
            t2.Start();
            Thread.Sleep(500);
        }

        private void Test_Client_Zero_Time_Rendezvous()
        {
            RendezvousChannel<int, int> rendChannel = new RendezvousChannel<int, int>();
            Thread t1 = new Thread(() => ClientFunc(1, rendChannel, 30, 0));
            Thread t2 = new Thread(() => ServerFunc(1, rendChannel, 5000, 500));

            t1.Start();
            Thread.Sleep(500);
            t2.Start();
            Thread.Sleep(500);
        }

        private void Test_Client_Zero_Time_After_Server_Starts_Rendezvous()
        {
            RendezvousChannel<int, int> rendChannel = new RendezvousChannel<int, int>();
            Thread t1 = new Thread(() => ServerFunc(1, rendChannel, 5000, 500));
            Thread t2 = new Thread(() => ClientFunc(1, rendChannel, 21, 0));

            t1.Start();
            Thread.Sleep(500);
            t2.Start();
            Thread.Sleep(500);
        }

        private void Test_Server_Zero_Time_Rendezvous()
        {
            RendezvousChannel<int, int> rendChannel = new RendezvousChannel<int, int>();
            Thread t1 = new Thread(() => ServerFunc(1, rendChannel, 0, 500));
            Thread t2 = new Thread(() => ClientFunc(1, rendChannel, 30, 5000));

            t1.Start();
            Thread.Sleep(500);
            t2.Start();
            Thread.Sleep(500);
        }

        private void Test_Server_Zero_Time_After_CLient_Starts_Rendezvous()
        {
            RendezvousChannel<int, int> rendChannel = new RendezvousChannel<int, int>();
            Thread t1 = new Thread(() => ClientFunc(1, rendChannel, 7, 5000));
            Thread t2 = new Thread(() => ServerFunc(1, rendChannel, 0, 500));

            t1.Start();
            Thread.Sleep(500);
            t2.Start();
            Thread.Sleep(500);
        }

        private void Test_Multiple_Client_One_Server_Rendezvous()
        {
            RendezvousChannel<int, int> rendChannel = new RendezvousChannel<int, int>();
            Thread t1 = new Thread(() => ClientFunc(1, rendChannel, 30, 5000));
            Thread t2 = new Thread(() => ClientFunc(2, rendChannel, 60, 5000));
            Thread t3 = new Thread(() => ServerFunc(1, rendChannel, 5000, 500));

            t1.Start();
            Thread.Sleep(500);
            t2.Start();
            Thread.Sleep(500);
            t3.Start();
            Thread.Sleep(500);
        }

        private void Test_Multiple_Server_One_Client_Rendezvous()
        {
            RendezvousChannel<int, int> rendChannel = new RendezvousChannel<int, int>();
            Thread t1 = new Thread(() => ClientFunc(1, rendChannel, 30, 5000));
            Thread t2 = new Thread(() => ServerFunc(2, rendChannel, 5000, 5000));
            Thread t3 = new Thread(() => ServerFunc(1, rendChannel, 5000, 500));

            t1.Start();
            Thread.Sleep(500);
            t2.Start();
            Thread.Sleep(500);
            t3.Start();
            Thread.Sleep(500);
        }

        private void Test_Multiple_Clients_First_And_Servers()
        {
            RendezvousChannel<int, int> rendChannel = new RendezvousChannel<int, int>();
            Thread t1 = new Thread(() => ClientFunc(1, rendChannel, 84, 5000));
            Thread t2 = new Thread(() => ClientFunc(2, rendChannel, 53, 5000));
            Thread t3 = new Thread(() => ServerFunc(2, rendChannel, 5000, 5000));
            Thread t4 = new Thread(() => ServerFunc(1, rendChannel, 5000, 500));

            t1.Start();
            Thread.Sleep(500);
            t2.Start();
            Thread.Sleep(500);
            t3.Start();
            Thread.Sleep(500);
            t4.Start();
            Thread.Sleep(500);
        }

        private void Test_Multiple_Servers_First_And_Clients()
        {
            RendezvousChannel<int, int> rendChannel = new RendezvousChannel<int, int>();
            Thread t1 = new Thread(() => ServerFunc(1, rendChannel, 5000, 500));
            Thread t2 = new Thread(() => ServerFunc(2, rendChannel, 5000, 500));
            Thread t3 = new Thread(() => ClientFunc(2, rendChannel, 53, 5000));
            Thread t4 = new Thread(() => ClientFunc(1, rendChannel, 84, 5000));

            t1.Start();
            Thread.Sleep(500);
            t2.Start();
            Thread.Sleep(500);
            t3.Start();
            Thread.Sleep(500);
            t4.Start();
            Thread.Sleep(500);
        }

        private void Test_Long_Wait_In_Rendezvous()
        {
            RendezvousChannel<int, int> rendChannel = new RendezvousChannel<int, int>();
            Thread t2 = new Thread(() => ServerFunc(1, rendChannel, 5000, 10000));
            Thread t1 = new Thread(() => ClientFunc(1, rendChannel, 10, 5000));

            t1.Start();
            Thread.Sleep(500);
            t2.Start();
            Thread.Sleep(500);
        }

        private void Test_Client_Interrupted_Rendezvous()
        {
            RendezvousChannel<int, int> rendChannel = new RendezvousChannel<int, int>();
            Thread t1 = new Thread(() => ClientFunc(1, rendChannel, 50, 5000));
            Thread t2 = new Thread(() => ClientFunc(2, rendChannel, 91, 5000));
            Thread t3 = new Thread(() => ServerFunc(1, rendChannel, 5000, 500));

            t1.Start();
            Thread.Sleep(500);
            t1.Interrupt();
            t2.Start();
            Thread.Sleep(500);
            t3.Start();
            Thread.Sleep(500);
        }

        private void Test_Server_Interrupted_Rendezvous()
        {
           RendezvousChannel<int, int> rendChannel = new RendezvousChannel<int, int>();
            Thread t1 = new Thread(() => ServerFunc(1, rendChannel, 5000, 500));
            Thread t2 = new Thread(() => ServerFunc(2, rendChannel, 5000, 500));
            Thread t3 = new Thread(() => ClientFunc(1, rendChannel, 10, 5000));

            t1.Start();
            Thread.Sleep(500);
            t1.Interrupt();
            t2.Start();
            Thread.Sleep(500);
            t3.Start();
            Thread.Sleep(500);
        }



        public static void ServerFunc(int id, RendezvousChannel<int, int> rendChannel, int timeout, int waitForReply)
        {
            try
            {
                Token<int> token = rendChannel.Accept(timeout);
                Console.WriteLine("Server "+ id +" started Rendezvous");
                Thread.Sleep(waitForReply);
                rendChannel.Reply(token,token.Value() *(-1));
            }
            catch (TimeoutException)
            {
                Console.WriteLine("Timeout exceeded in Server");
            }
            catch (ThreadInterruptedException)
            {
                Console.WriteLine("Server thread interrupted");
            }
        }
        public static void ClientFunc(int id, RendezvousChannel<int, int> rendChannel, int request, int timeout)
        {
            try
            {
                int response = rendChannel.Request(request,timeout);
                Console.WriteLine("Client "+ id + " asked double of " + request + " and got " + response);
            }
            catch (TimeoutException)
            {
                Console.WriteLine("Timeout exceeded in Client " + id);
            }
            catch (ThreadInterruptedException)
            {
                Console.WriteLine("Client thread interrupted");
            }
        }
    }
}

File: Java/RendezvousChannelSynchronizer/src/test/java/TestRendezvousChannel.java
import org.junit.Test;

/**
 * Created by Ruben Gomes on 24/05/2015.
 */
public class TestRendezvousChannel {


    //  //Responses -> Client gives an integer, server gives back the double of it.
    @Test
    public void test_server_starts_first() throws InterruptedException {
        RendezvousChannel<Integer,Integer> rendChannel = new RendezvousChannel<Integer,Integer>();
        Thread t1 = new ServerThread(1, rendChannel, 5000, 500);
        Thread t2 = new ClientThread(1, rendChannel,10,
                5000);

        t1.start();
        Thread.sleep(500);
        t2.start();
        Thread.sleep(1000);

    }

    @Test
    public void test_client_starts_first() throws InterruptedException {
        RendezvousChannel<Integer,Integer> rendChannel = new RendezvousChannel<Integer,Integer>();
        Thread t1 = new ClientThread(1, rendChannel,20,
                5000);
        Thread t2 = new ServerThread(1, rendChannel, 5000, 500);

        t1.start();
        Thread.sleep(500);
        t2.start();
        Thread.sleep(500);
    }
    @Test
    public void test_long_waiting_in_rendezvous()throws InterruptedException {
        RendezvousChannel<Integer,Integer> rendChannel = new RendezvousChannel<Integer,Integer>();
        Thread t2= new ServerThread(1, rendChannel, 5000, 5000);
        Thread t1 = new ClientThread(1, rendChannel,5,
                5000);

        t1.start();
        Thread.sleep(500);
        t2.start();
        Thread.sleep(5000);
    }

    @Test
    public void test_client_zero_time() throws InterruptedException {
        RendezvousChannel<Integer,Integer> rendChannel = new RendezvousChannel<Integer,Integer>();
        Thread t1 = new ClientThread(1, rendChannel,4, 0);
        Thread t2 = new ServerThread(1, rendChannel, 5000, 500);

        t1.start();
        Thread.sleep(500);
        t2.start();
        Thread.sleep(500);

    }

    @Test
    public void test_client_zero_time_after_server_starts()throws InterruptedException {
        RendezvousChannel<Integer,Integer> rendChannel = new RendezvousChannel<Integer,Integer>();
        Thread t1 = new ServerThread(1, rendChannel, 5000, 500);
        Thread t2 = new ClientThread(1, rendChannel,10, 0);

        t1.start();
        Thread.sleep(500);
        t2.start();
        Thread.sleep(500);

    }

    @Test
    public void test_server_zero_time() throws InterruptedException {
        RendezvousChannel<Integer,Integer> rendChannel = new RendezvousChannel<Integer,Integer>();
        Thread t1 = new ServerThread(1, rendChannel, 0, 500);
        Thread t2 = new ClientThread(1, rendChannel,10, 5000);

        t1.start();
        Thread.sleep(500);
        t2.start();
        Thread.sleep(500);
    }

    @Test
    public void test_server_zero_time_after_client_starts()throws InterruptedException {
        RendezvousChannel<Integer,Integer> rendChannel = new RendezvousChannel<Integer,Integer>();
        Thread t1 = new ServerThread(1, rendChannel, 5000, 500);
        Thread t2 = new ClientThread(1, rendChannel,10, 5000);

        t1.start();
        Thread.sleep(500);
        t2.start();
        Thread.sleep(1000);

    }

    @Test
    public void test_multiple_client_one_server()throws InterruptedException {
        RendezvousChannel<Integer,Integer> rendChannel = new RendezvousChannel<Integer,Integer>();
        Thread t1 = new ServerThread(1, rendChannel, 5000, 500);
        Thread t2 = new ClientThread(1, rendChannel,10, 5000);
        Thread t3 = new ClientThread(2, rendChannel,20, 5000);

        t1.start();
        Thread.sleep(500);
        t2.start();
        Thread.sleep(500);
        t3.start();
        Thread.sleep(1000);

    }

    @Test
    public void test_multiple_server_one_client()throws InterruptedException {
        RendezvousChannel<Integer,Integer> rendChannel = new RendezvousChannel<Integer,Integer>();
        Thread t1 = new ServerThread(1, rendChannel, 5000, 500);
        Thread t2 = new ServerThread(2, rendChannel, 5000, 500);
        Thread t3 = new ClientThread(1, rendChannel, 10, 5000);

        t1.start();
        Thread.sleep(500);
        t2.start();
        Thread.sleep(500);
        t3.start();
        Thread.sleep(1000);

    }

    @Test
    public void test_multiple_clients_first_and_servers()throws InterruptedException {
        RendezvousChannel<Integer,Integer> rendChannel = new RendezvousChannel<Integer,Integer>();
        Thread t1 = new ClientThread(1, rendChannel,10, 5000);
        Thread t2 = new ClientThread(2, rendChannel,20, 5000);
        Thread t3 = new ServerThread(1, rendChannel, 5000, 500);
        Thread t4 = new ServerThread(2, rendChannel, 5000, 500);

        t1.start();
        Thread.sleep(500);
        t2.start();
        Thread.sleep(500);
        t3.start();
        Thread.sleep(500);
        t4.start();
        Thread.sleep(500);

    }


    @Test
    public void test_multiple_servers_first_and_clients()throws InterruptedException {
        RendezvousChannel<Integer,Integer> rendChannel = new RendezvousChannel<Integer,Integer>();
        Thread t1 = new ServerThread(1, rendChannel, 5000, 500);
        Thread t2 = new ServerThread(2, rendChannel, 5000, 500);
        Thread t3 = new ClientThread(1, rendChannel,10, 5000);
        Thread t4 = new ClientThread(2, rendChannel,20, 5000);

        t1.start();
        Thread.sleep(500);
        t2.start();
        Thread.sleep(500);
        t3.start();
        Thread.sleep(500);
        t4.start();
        Thread.sleep(500);

    }

    @Test
    public void test_client_interrupted() throws InterruptedException {
        RendezvousChannel<Integer,Integer> rendChannel = new RendezvousChannel<Integer,Integer>();
        Thread t1 = new ClientThread(1, rendChannel,10, 5000);
        Thread t2 = new ClientThread(2, rendChannel,20, 5000);
        Thread t3 = new ServerThread(1, rendChannel,5000, 500);

        t1.start();
        Thread.sleep(500);
        t1.interrupt();
        t2.start();
        Thread.sleep(500);
        t3.start();
        Thread.sleep(800);
    }

    @Test
    public void test_server_interrupted() throws InterruptedException {
        RendezvousChannel<Integer,Integer> rendChannel = new RendezvousChannel<Integer,Integer>();
        Thread t1 = new ServerThread(1, rendChannel, 5000, 500);
        Thread t2 = new ServerThread(2, rendChannel, 5000, 500);
        Thread t3 = new ClientThread(1, rendChannel,20, 5000);

        t1.start();
        Thread.sleep(500);
        t1.interrupt();
        t2.start();
        Thread.sleep(500);
        t3.start();
        Thread.sleep(800);
    }


}
